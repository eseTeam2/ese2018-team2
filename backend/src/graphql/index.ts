import Mutation from "./resolvers/_mutations";
import Query from "./resolvers/_queries";

const resolvers = {
  Query,
  Mutation
}

export default resolvers;
