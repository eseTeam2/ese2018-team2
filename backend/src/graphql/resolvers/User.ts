import { UserResolvers } from "../__generated__/graphqlgen";

export const User: UserResolvers.Type = {
  ...UserResolvers.defaultResolvers,
  hasOrganizations: ({ id }, args, { userRepository, session }) =>
    userRepository.hasUserOrganization(id, session)
};

export default User;
