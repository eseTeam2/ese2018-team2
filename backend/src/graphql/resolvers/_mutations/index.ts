import { MutationResolvers } from "../../__generated__/graphqlgen";
import createJob from "./createJob";
import createOrganization from "./createOrganization";
import deleteJob from "./deleteJob";
import login from "./login";
import logout from "./logout";
import updateJob from "./updateJob";
import bookmarkJob from "./bookmarkJob";
import unbookmarkJob from "./unbookmarkJob";

export const Mutation: MutationResolvers.Type = {
  createJob,
  updateJob,
  deleteJob,
  createOrganization,
  login,
  logout,
  bookmarkJob,
  unbookmarkJob
};

export default Mutation;
