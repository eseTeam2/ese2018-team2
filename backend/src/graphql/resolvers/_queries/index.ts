import { QueryResolvers } from "../../__generated__/graphqlgen";
import jobs from "./jobs";
import me from "./me";
import organizations from "./organizations";
import applications from "./applications";
import roles from "./roles";
import users from "./users";
import search from "./search";
import searchAutocompletions from "./searchAutocompletions";

export const Query: QueryResolvers.Type = {
  jobs,
  roles,
  organizations,
  me,
  applications,
  users,
  search,
  searchAutocompletions
};

export default Query;
