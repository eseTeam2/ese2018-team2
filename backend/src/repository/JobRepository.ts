import { JobConnection } from "src/types";
import { Connection, Repository, Any } from "typeorm";
import { Job } from "../entity/Job";
import { Organization } from "../entity/Organization";
import { elasticClient } from "../lib/elastic";
import { Role } from "../entity/Role";

export interface JobUpdateArgs {
  id: string;
  title: string;
  description: string;
}

interface CursorOpts {
  seq: number;
  count: number;
  edge: string;
}

const encodeCursor = (options: string) =>
  Buffer.from(options).toString("base64");
const decodeCursor = (src: string) => Buffer.from(src, "base64").toString();

export class JobRepository {
  private connection: Connection;
  private jobs: Repository<Job>;
  private organizations: Repository<Organization>;

  constructor(connection: Connection) {
    this.connection = connection;
    this.jobs = connection.getRepository(Job);
    this.organizations = connection.getRepository(Organization);
  }

  // TODO create interface for argument type
  async getJobs(
    first: number,
    last: number,
    after: string,
    before: string
  ): Promise<JobConnection> {
    first = first || 10;

    const pageSize = first || last || 10;

    const getJobsAfter = async (job: string, howManyJobs: number) => {
      if (howManyJobs < 0) {
        throw new Error("first has to be positive");
      }

      const sequenceOfAfterEdge = (await this.jobs.findOne(job)).sequenceNumber;

      return this.jobs
        .createQueryBuilder("job")
        .limit(howManyJobs)
        .addOrderBy("job.sequenceNumber")
        .where("job.sequenceNumber > :seq", { seq: sequenceOfAfterEdge })
        .getMany();
    };

    const hasNextPage = async (job: string, howManyJobs: number) => {
      if (howManyJobs < 0) {
        throw new Error("first has to be positive");
      }

      const sequenceOfAfterEdge = (await this.jobs.findOne(job)).sequenceNumber;

      return (
        (await this.jobs
          .createQueryBuilder("job")
          .limit(howManyJobs)
          .addOrderBy("job.sequenceNumber")
          .where("job.sequenceNumber > :seq", { seq: sequenceOfAfterEdge })
          .getCount()) > 0
      );
    };

    const getJobsBefore = async (job: string, howManyJobs: number) => {
      if (howManyJobs < 0) {
        throw new Error("last has to be positive");
      }

      const sequenceOfBeforeEdge = (await this.jobs.findOne(job))
        .sequenceNumber;

      const allEdgesBefore = await this.jobs
        .createQueryBuilder("job")
        .addOrderBy("job.sequenceNumber")
        .where("job.sequenceNumber < :seq", { seq: sequenceOfBeforeEdge })
        .getCount();

      console.log(howManyJobs);

      // calculate the offset, how many rows need to be skipped until amount equals last
      const offset =
        allEdgesBefore - howManyJobs < 0 ? 0 : allEdgesBefore - howManyJobs;

      return this.jobs
        .createQueryBuilder("job")
        .offset(offset)
        .addOrderBy("job.sequenceNumber")
        .where("job.sequenceNumber < :seq", { seq: sequenceOfBeforeEdge })
        .getMany();
    };

    const hasPreviousPage = async (job: string, howManyJobs: number) => {
      if (howManyJobs < 0) {
        throw new Error("last has to be positive");
      }

      const sequenceOfBeforeEdge = (await this.jobs.findOne(job))
        .sequenceNumber;

      const allEdgesBefore = await this.jobs
        .createQueryBuilder("job")
        .addOrderBy("job.sequenceNumber")
        .where("job.sequenceNumber < :seq", { seq: sequenceOfBeforeEdge })
        .getCount();

      return allEdgesBefore - howManyJobs >= 0;
    };

    let result: Array<Job> = [];

    // check if cursor exists
    if (
      (before || after) && (await this.jobs.findByIds([decodeCursor(after || before)])).length === 0
    ) {
      return {
        nodes: result,
        pageInfo: {
          endCursor: null,
          hasNextPage: false,
          hasPreviousPage: false,
          startCursor: null
        },
        totalCount: await this.jobs.count()
      };
    }

    if (after) {
      result = await getJobsAfter(decodeCursor(after), first);
    }

    if (before) {
      result = await getJobsBefore(decodeCursor(before), last);
    }

    if (!after && !before) {
      result = await this.jobs
        .createQueryBuilder("job")
        .limit(first)
        .addOrderBy("job.sequenceNumber")
        .getMany();
    }

    const next = await hasNextPage(result[result.length - 1].id, pageSize);
    const previous = await hasPreviousPage(result[0].id, pageSize);

    const nodes = await this.jobs.findByIds(result.map(e => e.id));

    return {
      nodes,
      pageInfo: {
        endCursor: next ? encodeCursor(nodes[nodes.length - 1].id) : null,
        hasNextPage: next,
        hasPreviousPage: previous,
        startCursor: previous ? encodeCursor(nodes[0].id) : null
      },
      totalCount: await this.jobs.count()
    };
  }

  async createJob(args: any): Promise<Job> {
    const job = new Job();
    job.title = args.input.title;
    job.description = args.input.description;
    job.salary = 1.0;
    job.start = new Date();

    const organization = await this.organizations.findOneOrFail(
      args.input.organization
    );
    job.organization = organization;
    // save
    await this.jobs.insert(job);
    return job;
  }

  async deleteJob(id: string) {
    await this.jobs.delete(id);
    return true;
  }

  async updateJob(args: JobUpdateArgs): Promise<Job> {
    const id = Object.entries(args)
      .filter(e => e[0] === "userId")
      .map(e => e[1])[0];

    const fieldsToUpdate = Object.entries(args)
      .filter(e => e[0] !== "id")
      .reduce(
        (prev, curr) => Object.assign({}, prev, { [curr[0]]: curr[1] }),
        {}
      );

    await this.jobs.update({ id }, fieldsToUpdate);

    return this.jobs.findOneOrFail(id);
  }

  async search(minSalary: number, maxSalary: number) {

    await elasticClient.ping({
      requestTimeout: 30000
    });

    const searchResult = await elasticClient.search({
      index: "jobs",
      body: {
        query: {
          range: {
            salary: {
              gte: minSalary,
              lte: maxSalary
            }
          }
        },
        aggs: {
          Job: {
            terms: {
              field: "roles"
            }
          }
        }
      }
    });

    const ids = searchResult.hits.hits.map((e) => e._id)
    const nodes = await this.jobs.findByIds(ids)

    const todo = searchResult.aggregations["Job"].buckets.map((e:any) => ({key: parseInt(e.key), count: e.doc_count}))
    console.log(todo)

    const roles = await this.connection.getRepository(Role).find()

    // THIS IS EVIL >:D
    const buckets = todo.map(async (bucket:any) => {

      const role = (await this.connection.getRepository(Role).find({ sequenceNumber: bucket.key }))[0]
      
      return {
        role,
        count: bucket.count
      }
    });

    console.log(`Buckets: ${JSON.stringify(buckets)}`)

    return {
      nodes,
      buckets
    }
  }
}

export default JobRepository;
