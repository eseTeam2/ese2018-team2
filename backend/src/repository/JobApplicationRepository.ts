import { Connection, Repository } from "typeorm";
import { JobApplication } from "../entity/JobApplication";
import Utils, { getUserId } from "./Utils";
import { Job } from "../entity/Job";
import { User } from "../entity/User";
import JobApplicationState from "../entity/JobApplicationState";
import enforceAuth from "./Utils";

export class JobApplicationRepository {
  private connection: Connection;
  private applications: Repository<JobApplication>;
  private jobs: Repository<Job>;
  private users: Repository<User>;

  constructor(connection: Connection) {
    this.connection = connection;
    this.applications = connection.getRepository(JobApplication);
    this.jobs = connection.getRepository(Job);
    this.users = connection.getRepository(User);
  }

  getApplications(
    args: any,
    session: Express.Session
  ): Promise<JobApplication[]> {
    enforceAuth(session);

    return this.applications
      .createQueryBuilder("applications")
      .leftJoinAndSelect("applications.job", "jobs")
      .leftJoinAndSelect("applications.user", "users")
      .getMany();
  }

  async applyForJob(jobId: string, session: Express.Session): Promise<any> {
    enforceAuth(session);

    const application = new JobApplication();
    application.job = await this.jobs.findOneOrFail(jobId);
    application.user = await this.users.findOneOrFail(session.user.id);
    await this.applications.insert(application);
    console.log(application);
    return true;
  }

  async approveJobApplication(
    applicationId: string,
    session: Express.Session
  ): Promise<any> {
    enforceAuth(session);

    await this.applications.update(
      { id: applicationId },
      { state: JobApplicationState.APPROVED }
    );

    return true;
  }

  async rejectJobApplication(
    applicationId: string,
    session: Express.Session
  ): Promise<any> {
    enforceAuth(session);

    await this.applications.update(
      { id: applicationId },
      { state: JobApplicationState.REJECTED }
    );

    return true;
  }

  /**
   * Check if current user already applied to a specific job
   * @param jobId Job to check
   * @param session 
   */
  async isApplied(jobId: string, session: Express.Session): Promise<boolean> {
    // If userId is not defined, immediatly return false
    // This only works, when
    if (!session) {
      return false;
    }

    const result = await this.applications
      .createQueryBuilder("jobApplications")
      .where('"userId" = :userId', { userId: getUserId(session) })
      .andWhere('"jobId" = :jobId', { jobId })
      .getOne();

    return result != null;
  }
}
