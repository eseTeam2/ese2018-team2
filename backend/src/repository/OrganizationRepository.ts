import { Connection, Repository } from "typeorm";
import { Job } from "../entity/Job";
import { Organization } from "../entity/Organization";
import { User } from "../entity/User";
import enforceAuth, { enforceAdmin, getUserId, isAdmin } from "./Utils";

export class OrganizationRepository {
  private connection: Connection;
  private organizations: Repository<Organization>;
  private users: Repository<User>;
  private jobs: Repository<Job>;

  constructor(connection: Connection) {
    this.connection = connection;
    this.organizations = connection.getRepository(Organization);
    this.users = connection.getRepository(User);
    this.jobs = connection.getRepository(Job);
  }

  async getOrganizations(
    session: Express.Session,
    organizationId?: string
  ): Promise<Organization[]> {
    enforceAuth(session);
    if (isAdmin(session)) {
      if (organizationId) {
        return this.organizations.findByIds([organizationId]);
      }
      return this.organizations.find();
    }

    //TODO check if id is an organization of the current user (organizational user)
    //TODO if user is not a organization user, then do not load!!!
    return (await this.users.findByIds([getUserId(session)], {
      relations: ["employer"]
    }))[0].employer;
  }

  async createOrganization(name: string): Promise<Organization> {
    const organization = new Organization();
    organization.name = name;
    await this.organizations.insert(organization);
    return organization;
  }

  async approveOrganization(
    organizationId: string,
    session: Express.Session
  ): Promise<any> {
    enforceAdmin(session);

    await this.organizations.update({ id: organizationId }, { approved: true });
    return true;
  }

  async deleteOrganization(organizationId: string) {
    await this.jobs.delete({ organization: { id: organizationId } });
    await this.organizations.delete({ id: organizationId });
    return true;
  }
}
