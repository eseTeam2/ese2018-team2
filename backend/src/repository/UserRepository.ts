import {
  Connection,
  getConnection,
  Repository,
  FindManyOptions
} from "typeorm";
import bcrypt from "bcryptjs";
import { User } from "../entity/User";
import { Job } from "../entity/Job";
import Utils, { enforceAdmin } from "./Utils";
import enforceAuth from "./Utils";

export interface FindUserOptions {
  onlyAdmins?: boolean;
  id?: string;
}

export class UserRepository {
  private users: Repository<User>;
  private jobs: Repository<Job>;

  constructor(connection: Connection) {
    this.users = connection.getRepository(User);
    this.jobs = connection.getRepository(Job);
  }

  async login(email: string, password: string, session: Express.Session) {
    const foundUsers = await this.users.find({ email });
    if (foundUsers.length === 0) {
      return false;
    }

    const user = foundUsers[0];
    if (!bcrypt.compareSync(password, user.password)) {
      return false;
    }

    session.user = user;

    return true;
  }

  async findUsers(
    { onlyAdmins, id }: FindUserOptions,
    session: Express.Session
  ) {
    enforceAdmin(session);

    // search by id
    if (id) {
      return this.users.find({ id });
    }

    // TODO won't work for more where
    const whereOnlyAdmins = (
      obj: FindManyOptions<User>,
      siteAdmin?: boolean
    ): FindManyOptions<User> =>
      siteAdmin ? { ...obj, where: { siteAdmin: true } } : obj;

    // create sort in where
    const order = (obj: FindManyOptions<User>): FindManyOptions<User> => ({
      ...obj,
      order: { sequenceNumber: "ASC" }
    });

    return this.users.find(whereOnlyAdmins(order({}), onlyAdmins));
  }

  async getMe(session: Express.Session) {
    enforceAuth(session);
    return this.users.findOneOrFail(session.user.id);
  }

  async addBookmark(jobId: string, session: Express.Session): Promise<any> {
    enforceAuth(session);

    await getConnection()
      .createQueryBuilder()
      .relation(User, "bookmarkedJobs")
      .of(session.user.id)
      .add(jobId);

    return true;
  }

  async removeBookmark(jobId: string, session: Express.Session): Promise<any> {
    enforceAuth(session);

    await getConnection()
      .createQueryBuilder()
      .relation(User, "bookmarkedJobs")
      .of(session.user.id)
      .remove(jobId);

    return true;
  }
}
