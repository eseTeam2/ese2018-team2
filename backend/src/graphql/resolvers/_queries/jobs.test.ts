import { importSchema } from "graphql-import";
import { graphql } from "graphql";
import resolvers from "../../";
import { makeExecutableSchema } from "graphql-tools";
import { Context, Job } from "../../../types";
import JobRepository from "../../../repository/JobRepository";
import { OrganizationRepository } from "../../../repository/OrganizationRepository";
import { UserRepository } from "../../../repository/UserRepository";
import { JobApplicationRepository } from "../../../repository/JobApplicationRepository";
import { SkillRepository } from "../../../repository/SkillRepository";

describe("Get likedJobs", () => {
  const schema = makeExecutableSchema({
    typeDefs: importSchema("./schema.graphql"),
    resolvers
  } as any);

  it("should return an empty list", async () => {
    const mockCtx: Context = {
      jobRepository: new (jest.fn<JobRepository>(() => ({
        getJobs: () => []
      })))(),
      organizationRepository: new (jest.fn<OrganizationRepository>())(),
      userRepository: new (jest.fn<UserRepository>())(),
      applicationRepository: new (jest.fn<JobApplicationRepository>())(),
      skillRepository: new (jest.fn<SkillRepository>())()
    };

    const query = `
      query {
        jobs (first: 1) {
          nodes: {
            id
          }
        }
      }
    `;
    expect(await graphql(schema, query, null, mockCtx)).toMatchSnapshot();
  });
});
