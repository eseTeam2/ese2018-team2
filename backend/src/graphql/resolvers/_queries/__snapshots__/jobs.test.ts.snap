// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get jobs should return a job 1`] = `
Object {
  "data": Object {
    "jobs": Array [
      Object {
        "id": "123",
        "organization": Object {
          "id": "123",
          "name": "Hello",
        },
        "title": "Hello world",
      },
    ],
  },
}
`;

exports[`Get jobs should return an empty list 1`] = `
Object {
  "data": Object {
    "jobs": Array [],
  },
}
`;
